### ISSUE TITLE
Ability to transform all configuration

### ISSUE THREAD

**Issue by TPGEric** [Microsoft/slow-cheetah#145]
Hi, 

I need to transform all configuration to output directory on build (debug, staging, release).
How can I achieve this with 3.1 ?
My deployment tool will catch the correct config file depending on release process.

Thanks in advance.

**Comment by davilimap** [comment]
I'm not sure I understand what you need. Are you building all configurations at once or just one configuration, but you need all the transformations to occur? In the second case, would the transformations be done to one folder or to the respective configuration folders?

**Comment by TPGEric** [comment]
My need is to generate all transformations on one build into the current configuration destination folder.
Ie: if I build a release configuration :
- MyApplication.exe.Config (Transformed for release configuration)
- MyApplication.Debug.Config (Transformed from debug configuration)
- MyApplication.Staging.Config (Transformed from statging configuration)
- MyApplication.Release.config (Transformed from release configuration)

**Comment by davilimap** [comment]
This is not included in SlowCheetah's main functionality, but can be achieved through MSBuild targets. See the [docs](https://github.com/Microsoft/slow-cheetah/blob/master/doc/targets.md) for information on how the SlowCheetah targets work (also [here](https://msdn.microsoft.com/en-us/library/ms171462.aspx) for more info on targets). It seems like you can just add your files to the list of files that are transformed, **ScFilesToTransform**.

**Comment by MikaelUmaN** [comment]
Hi @davilimap . I believe I have a similar situation. When performing a Release build, I want to transform the original App.config file into X different environments. All should be built according to the release configuration - I don't want to make X different compilations just because I need X different app config files.

Is that possible to achieve? I tried something like:

```
  <Target Name="ScApplyTransforms">
	<ItemGroup>
		<ScFilesToTransform Include="@(None);@(Content);@(Resource);@(EmbeddedResource)" Condition=" '%(TransformOnBuild)' == 'true' ">
			<SourceFile>App.config</SourceFile>
			<TransformFile>App.QA.config</TransformFile>
			<DestinationFile>App.QA.trans.config</DestinationFile>
		</ScFilesToTransform>
	</ItemGroup>
  </Target>
```

based on what I could find on the web and your links but I'm not having any luck. The regular transforms (e.g. App.Release.config and App.Debug.config) work without any problem.

Basically I just want extra config files for my release build, with differing connection strings and app specific settings etc. Is that achievable?

**Comment by wenlin1991** [comment]
> This is not included in SlowCheetah's main functionality, but can be achieved through MSBuild targets. See the [docs](https://github.com/Microsoft/slow-cheetah/blob/master/doc/targets.md) for information on how the SlowCheetah targets work (also [here](https://msdn.microsoft.com/en-us/library/ms171462.aspx) for more info on targets). It seems like you can just add your files to the list of files that are transformed, **ScFilesToTransform**.

Hi @davilimap , can we expect this as a new feature in next version?

**Comment by SHAGGAR** [comment]
I found myself wanting this same thing and after flailing around with it for a while (i hate msbuild) I ended up with something that works.  There may be a way better way to do this, but it worked for me.  I put this at the end of my csproj.
```
<ItemDefinitionGroup>
	<None>
		<IsTransformFile>false</IsTransformFile>
		<TransformOnBuild>false</TransformOnBuild>
	</None>
</ItemDefinitionGroup>
<Target Name="TransformAllConfigs" DependsOnTargets="PrepareForBuild;AssignTargetPaths;ScCollectTransformFiles;ScTransformAppConfig">
	<ItemGroup>
		<_TransformFiles Include="@(_NoneWithTargetPath->'%(FullPath)')" Condition="'%(IsTransformFile)' == 'true'"/>
	</ItemGroup>

	<PropertyGroup>
		<AllConfigOutDir Condition="'$(AllConfigOutDir)' == ''">$(OutDir)</AllConfigOutDir>
	</PropertyGroup>

	<MakeDir Directories="$(AllConfigOutDir)" Condition=" !Exists('$(AllConfigOutDir)') " />

	<SlowCheetah.TransformTask Source="$(ScIntermediateAppConfig)" 
		Transform="%(_TransformFiles.RelativeDir)%(_TransformFiles.Filename)%(_TransformFiles.Extension)" 
		Destination="$(AllConfigOutDir)%(_TransformFiles.Filename)%(_TransformFiles.Extension)"/>
</Target>
```
This adds a `TransformAllConfigs task which can be called with msbuild directly by doing.  My intention is to call it as a separate build task using msbuild (msbuild {project}.csproj /t:TransformAllConfigs /p:AllConfigOutDir=bin\config\) but you could add it to your AfterBuild target if you wanted to build all configs every time your project is built.

The AllConfigOutDir parameter is optional and lets you specify an output dir for just the transformed configs.  If you dont specify  AllConfigOutDir it will just use OutDir which will result in all the configs being put into the normal output with the rest of your build output.
